As a user, I expect an evaluation of RFC vs USGS that succeeds in production using 6.33 to succeed in 6.34 and 7.0

The declaration, with necessary omissions, is below. I'll gather the WRDS data,

Hank

===========

```yaml
label: Example Meal for 6-hour AHPS vs USGS
observed:
  label: USGS NWIS Instantaneous Streamflow Observations
  sources:
  - interface: usgs nwis
    uri: https://nwis.waterservices.usgs.gov/nwis/iv
  variable:
    name: '00060'
  type: observations
predicted:
  label: WRDS AHPS Streamflow Forecasts
  sources:
  - interface: wrds ahps
    uri: https://[omitted]/api/rfc_forecast/v2.0/forecast/streamflow
  variable:
    name: QR
unit: ft3/s
lead_time_pools:
  period: 6
  frequency: 6
  unit: hours
lead_times:
  minimum: 0
  maximum: 120
  unit: hours
reference_dates:
  minimum: '2025-08-24T00:00:00Z'
  maximum: '2025-09-03T15:51:37Z'
valid_dates:
  minimum: '2025-08-24T00:00:00Z'
  maximum: '2025-09-03T15:51:37Z'
unit_aliases:
- alias: herr
  unit: hank
features:
- observed: '01631000'
  predicted: FROV2
- observed: '01423000'
  predicted: WALN6
feature_service:
  uri: https://[omitted]/api/location/v3.0/metadata
metrics:
- name: root mean square error
- name: pearson correlation coefficient
- name: sample size
- name: volumetric efficiency
duration_format: hours
output_formats:
- format: csv2
- format: netcdf2
- format: png
- format: pairs
- format: protobuf
```

Sorry, was rushing. The error is the standard no data found error. Example:

```
2025-09-03T15:57:14.797+0000  [pool-23-thread-1] INFO EvaluationService - The evaluation failed with the following stack trace: 
wres.pipeline.InternalWresException: Could not complete project execution
	at wres.pipeline.Evaluator.evaluate(Evaluator.java:408)
	at wres.pipeline.Evaluator.evaluate(Evaluator.java:181)
	at wres.Functions.evaluate(Functions.java:131)
	at wres.server.EvaluationService.lambda$startEvaluation$2(EvaluationService.java:695)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:840)
Caused by: wres.pipeline.WresProcessingException: Encountered an error while processing evaluation 'BGoUlQWVglzLfDfwEzYlE4RYD2I': 
	at wres.pipeline.Evaluator.evaluate(Evaluator.java:855)
	at wres.pipeline.Evaluator.evaluate(Evaluator.java:379)
	... 7 common frames omitted
Caused by: wres.pipeline.WresProcessingException: Statistics could not be produced for any pools. This probably occurred because none of the pools contained valid pairs. Check that the declaration contains some pools whose boundaries (e.g., earliest and latest issued times, earliest and latest valid times and earliest and latest lead durations) are sufficiently broad to capture some pairs at the desired time scale for at least one feature and threshold.
	at wres.pipeline.pooling.PoolReporter.report(PoolReporter.java:230)
	at wres.pipeline.Evaluator.evaluate(Evaluator.java:798)
	... 8 common frames omitted
```

Hank

--- a/wres-io/src/wres/io/retrieving/database/TimeSeriesRetriever.java
+++ b/wres-io/src/wres/io/retrieving/database/TimeSeriesRetriever.java
@@ -26,6 +26,7 @@ import java.util.function.Function;
 import java.util.function.Supplier;
 import java.util.stream.Stream;
 
+import org.apache.commons.lang3.tuple.Pair;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
@@ -1178,43 +1179,23 @@ abstract class TimeSeriesRetriever<T> implements Retriever<TimeSeries<T>>
         // Get the valid time
         Instant validTime = provider.getInstant( "valid_time" );
 
-        Map<ReferenceTimeType, Instant> referenceTimes = Collections.emptyMap();
+        Pair<ReferenceTimeType, Map<ReferenceTimeType, Instant>>
+                referenceTimePair = this.getReferenceTimes( provider,
+                                                            lastBuilder );
 
-        // Add the explicit reference time
-        if ( provider.hasColumn( REFERENCE_TIME )
-             && !provider.isNull( REFERENCE_TIME ) )
+        Map<ReferenceTimeType, Instant> referenceTimes = referenceTimePair.getValue();
+        ReferenceTimeType currentReferenceTimeType = referenceTimePair.getKey();
+
+        // Only add events whose valid times can be properly deduced from the reference time and lead time. For example,
+        // when the reference time is a ReferenceTimeType.GENERATION_TIME, the valid time cannot be derived from the
+        // reference time plus the lead duration. This is a flaw in the db schema that needs to be addressed, i.e.,
+        // store time-series by valid time, not lead duration.
+        if ( this.canDeduceValidTimeFromReferenceTimeAndLeadDuration( currentReferenceTimeType ) )
         {
-            ReferenceTimeType type = ReferenceTimeType.UNKNOWN;
-
-            if ( provider.hasColumn( REFERENCE_TIME_TYPE ) )
-            {
-                String typeString = provider.getString( REFERENCE_TIME_TYPE );
-                type = ReferenceTimeType.valueOf( typeString.toUpperCase() );
-            }
-
-            Instant referenceTime = provider.getInstant( REFERENCE_TIME );
-            referenceTimes = new EnumMap<>( ReferenceTimeType.class );
-            referenceTimes.put( type, referenceTime );
-            referenceTimes = Collections.unmodifiableMap( referenceTimes );
+            // Add the event
+            Event<S> event = mapper.apply( provider );
+            this.addEventToTimeSeries( event, lastBuilder );
         }
-
-        // Adding a new reference time to an existing series or adding events to a new series?
-        if ( this.hasNewReferenceTimes( lastBuilder, referenceTimes ) )
-        {
-            TimeSeriesMetadata oldMetadata = lastBuilder.getMetadata();
-            Map<ReferenceTimeType, Instant> aggregateReferenceTimes = new EnumMap<>( ReferenceTimeType.class );
-            aggregateReferenceTimes.putAll( oldMetadata.getReferenceTimes() );
-            aggregateReferenceTimes.putAll( referenceTimes );
-            TimeSeriesMetadata adjustedMetadata = new TimeSeriesMetadata.Builder( oldMetadata )
-                    .setReferenceTimes( aggregateReferenceTimes )
-                    .build();
-            lastBuilder.setMetadata( adjustedMetadata );
-            return Collections.unmodifiableList( returnMe );
-        }
-
-        // Add the event     
-        Event<S> event = mapper.apply( provider );
-        this.addEventToTimeSeries( event, lastBuilder );
 
         // Add the timescale info
         String functionString = provider.getString( "scale_function" );
@@ -1280,21 +1261,65 @@ abstract class TimeSeriesRetriever<T> implements Retriever<TimeSeries<T>>
     }
 
     /**
-     * @param builder the time-series builder
-     * @param <S> the time-series data type
-     * @return whether the builder contains reference times that do not match the new reference times
+     * @param type the reference time type
+     * @return whether the valid time is equal to the reference time plus the lead duration
      */
 
-    private <S> boolean hasNewReferenceTimes( TimeSeries.Builder<S> builder,
-                                              Map<ReferenceTimeType, Instant> newReferenceTimes )
+    private boolean canDeduceValidTimeFromReferenceTimeAndLeadDuration( ReferenceTimeType type )
     {
-        return Objects.nonNull( builder.getMetadata() )
-               && !builder.getMetadata()
-                          .getReferenceTimes()
-                          .isEmpty()
-               && !Objects.equals( builder.getMetadata()
-                                          .getReferenceTimes(),
-                                   newReferenceTimes );
+        return Objects.isNull( type )
+               || type == ReferenceTimeType.T0
+               || type == ReferenceTimeType.ISSUED_TIME
+               || type == ReferenceTimeType.ANALYSIS_START_TIME;
+    }
+
+    /**
+     * Obtains the reference times from the provider, including the type of the current reference time.
+     *
+     * @param <S> the time-series data type
+     * @param provider the data provider
+     * @param lastBuilder the last time-series builder
+     * @return the aggregate reference times from the current read and the existing builder
+     */
+
+    private <S> Pair<ReferenceTimeType, Map<ReferenceTimeType, Instant>> getReferenceTimes( DataProvider provider,
+                                                                                            TimeSeries.Builder<S> lastBuilder )
+    {
+        Map<ReferenceTimeType, Instant> referenceTimes = Collections.emptyMap();
+
+        ReferenceTimeType currentReferenceTimeType = null;
+
+        // Add the explicit reference time
+        if ( provider.hasColumn( REFERENCE_TIME )
+             && !provider.isNull( REFERENCE_TIME ) )
+        {
+            ReferenceTimeType type = ReferenceTimeType.UNKNOWN;
+
+            if ( provider.hasColumn( REFERENCE_TIME_TYPE ) )
+            {
+                String typeString = provider.getString( REFERENCE_TIME_TYPE );
+                type = ReferenceTimeType.valueOf( typeString.toUpperCase() );
+                currentReferenceTimeType = type;
+            }
+
+            Instant referenceTime = provider.getInstant( REFERENCE_TIME );
+            referenceTimes = new EnumMap<>( ReferenceTimeType.class );
+            referenceTimes.put( type, referenceTime );
+
+            // If the existing builder already has reference times, add those
+            if ( Objects.nonNull( lastBuilder.getMetadata() )
+                 && !lastBuilder.getMetadata()
+                                .getReferenceTimes()
+                                .isEmpty() )
+            {
+                referenceTimes.putAll( lastBuilder.getMetadata()
+                                                  .getReferenceTimes() );
+            }
+
+            referenceTimes = Collections.unmodifiableMap( referenceTimes );
+        }
+
+        return Pair.of( currentReferenceTimeType, referenceTimes );
     }
 
     /**

